ui.rb

```diff
@@ -0,0 +1,104 @@
+require 'singleton'
+require 'date'
+
+class UserInterface
+    include Singleton
+
+    def get_date
+        begin
+            print "Enter date (YYYY-MM-DD): "
+            date=Date.parse(gets.chomp)
+        rescue Date::Error
+            puts "Please enter valid date: "
+            retry
+        end
+
+        return date
+    end
+
+    def get_year
+        loop do
+            print "Enter year (YYYY): "
+            year = gets.chomp
+
+            if year.match?(/^\d{4}/)
+                return year.to_i
+            else
+                puts "Please enter valid year"
+            end
+        end
+    end
+
+    def get_month
+        loop do
+            print "Enter month (MM): "
+            month = gets.chomp
+
+            if (1..12).include?(month.to_i)
+                return month.to_i
+            else
+                puts "Please enter valid month"
+            end
+        end
+    end
+
+
+    
+    private def display_event(event)
+        puts "\nTitle: #{event.title}"
+        puts "Description: #{event.description}\n"
+        puts "-"*40
+    end
+
+    def list_events(events)
+        if !(events.empty?())
+            events.each do |event|
+                display_event(event)
+            end
+        else
+            puts "No event to show"
+        end
+    end
+
+    private def event_exists?(events,title)
+        events.any? {|event| event.title==title}
+    end
+
+    def get_event_title(events)
```

Comment:
method name should be comprehensive, it should convey proper meaning of the functionality.